### DISCUSSION TITLE
很少人注意到 Python 的進階語法

### DISCUSSION THREAD

**Discussion by 32844583** [1101-GS3073-C/Python-Community-GS3073-C#541] [Ideas]
### Decorator
某天無意間在網路上看到一教學影片教到 Python 的 decorator，個人看完主要就是在如何封裝 function、class，以下著重在 Decorator 如何使用、使用時機，Decorator 被大量廣泛的使用在各方 library/package 中，之後想寫 Python 或覺得這個語言可以撐很久的可以看看

https://medium.com/citycoddee/python%E9%80%B2%E9%9A%8E%E6%8A%80%E5%B7%A7-3-%E7%A5%9E%E5%A5%87%E5%8F%88%E7%BE%8E%E5%A5%BD%E7%9A%84-decorator-%E5%97%B7%E5%97%9A-6559edc87bc0

**Comment by BingRong1217** [comment]
這個網站寫得很好也對我很有幫助!!

**Comment by awkward-willy** [comment]
裝飾器在 Python 確實相當好用且強大，在實作 Flask 的過程中我們也常常會使用到，的確是需要時間精進的觀念之一。

**Comment by williamyen0623** [comment]
以前在學java似乎有學過很多類似的知識，但感覺python又不太一樣，把這個學好肯定能幫助我們把程式寫的更有架構、更有系統性。

**Comment by HsuChen88** [comment]
才知道原來上課用的@app.route('/')就是裝飾器啊

**Comment by skyling0299** [comment]
以前一直覺得@超級怪，為甚麼可以@什麼什麼
後來到了開始玩js跟架網站突然又覺得不怪了

**Comment by RogelioKG** [comment]
## Type-Hinting
我發現好像也蠻少人在使用 Python 的型別提示 (type-hinting)，
有寫過 C 或者 Java 這類靜態型別語言的朋友應該會很熟悉，
( Python 的型別提示是寫給人看的，
不會真正影響到底層資料結構的儲存方式，
也不會影響執行期的速度。
不過有一些型別檢查工具，例如 mypy，
可以利用這個語法在執行前進行型別檢查)
如果 IDE 是 VScode 的話，
只要在函數的參數加上型別提示，
一旦嘗試調用參數的 method
VScode 就會對根據型別提示給出可用 method 的提示，
並且 highlighter 也能夠正確上色。

**Reply by RogelioKG** [reply]
```py
" list "
a: list[int]
# 一個 list，裡面的元素都是 int
# a = [1, 2, 3, 4, 5]

" dict "
b: dict[str, int]
# 一個 dict，鍵值對是 str : int
# b = {"Jerry": 18, "Larry": 30}

" tuple "
c: tuple[str, int, float]
# 一個 tuple，裡面的元素依序是 str, int, float
# c = ("Jerry", 20, 168.5)
```

**Reply by RogelioKG** [reply]
```py
" | 二選一 "
e: list[int | float]
# 一個 list，裡面的元素是 int or float
# e = [1, 2, 3, 4, 5]
# e = [1.1, 2.2, 3.3, 4.4, 5.5]

" Callable 函數 "
func: Callable[[str, int], bool]
# 一個 function，接受兩個引數 str, int，回傳 bool
# def my_function(name: str, age: int) -> bool:
#     pass
# func = my_function
" Any 都可 "
var: list[Any]
# var = [1, "2", {3: "4"}, (5,)]
```
